### ISSUE TITLE
Idea: Preinstall "official" modules, activate through config

### ISSUE THREAD

**Issue by etiennedi** [semi-technologies/weaviate#1439]
## Current State / Idea

Previously the module design was mostly driven by the idea that "official" modules (meaning modules we publish under the name of SeMI) do not differ from community modules (meaning modules a user might build privately, but never publish). While I strongly believe this is a principle we have to keep, we can still make using the most common modules more convenient.

Currently Weaviate Core (without any modules) would literally ship without the code for any modules. The current process to enable a module is to "install" it. I.e. to copy in the module before starting Weaviate. To make the most common module combinations more convenient we are publishing (or planning on publishing) images, such as `weaviate-text`, where the officially supported text module (i.e. currently `text2vec-contextionary` is preinstalled and activated).

## Proposal

* We **decouple module installation and module activation**, i.e. even if 50 modules are installed, if none are activated, Weaviate is indistinguishable from having no modules installed
* This allows us to ship just a single image with the modules we officially support (i.e the ones we also advertise on the website)
* From a tech perspective, this does not mean that each module will be loaded, but only act depending on config. Weaviate Core will only load the modules which are activated, so that pre-shipping with many modules does not alter startup time, mem usage, etc., if they are not activated.

### Example API

The modules could simply be activated through configuration, e.g.
```
ENABLE_MODULES=text2vec-contextionary,img2vec-resnet,network
```

If the user uses the config gen on the website, they'll never have to make this setting themselves, they only need to select the appropriate option from the dropdown.

### Advantages
* We do not need to build a large number of combinations of images, such as `weaviate`, `weaviate-text`, `weaviate-image`, `weaviate-text-image`, `weaviate-text-image-network`, etc.
* Community modules are not affected by this change, they can still be copied in before starting Weaviate, the only difference is that some modules are already copied in

### Disadvantages
* The default Weaviate image will be bigger as we add more official images, the weight has to be carried around even if users only use a small portion of the modules. However, given that compiled Go code is not too large, this should not be a problem. If there is a situation where every MB matters, then a user could still build a custom image which starts from no modules and only adds in the ones they need.

Looking forward to you feedback!
cc @laura-ham @antas-marcin @bobvanluijt

**Comment by bobvanluijt** [comment]
Quick question: aren't the c11y modules quite big? I.e., they are in the gig size range, right?

For the rest, I can easily imagine this is significantly easier (for example on cloud marketplace deployments)

**Comment by etiennedi** [comment]
> Quick question: aren't the c11y modules quite big? I.e., they are in the gig size range, right?

Yes, but the contextionary model is decoupled (separate container) from the `text2vec-contextionary` module, so even if you have the `text2vec-contextionary` module pre-installed, does not mean you need to spin up the contextionary model container.

**Comment by bobvanluijt** [comment]
Gotcha - in that case (i.e., from a productization perspective) I’m all for

**Comment by laura-ham** [comment]
That looks like a very good idea!

Question to see if I understand it correctly; you mention:
* `ENABLE_MODULES=text2vec-contextionary,img2vec-resnet,network`: So for example if a user want to run `weaviate-text`, this line will just be `ENABLE_MODULES=text2vec-contextionary`? And selecting `weaviate-text-image` will be e.g. `ENABLE_MODULES=text2vec-contextionary,img2vec-resnet`? So instead of deploying an image per configuration we'll have one image with all modules and the activation of these modules is defined in de configuration / env var `ENABLE_MODULES`?

**Comment by etiennedi** [comment]
Absolutely correct. As Bob pointed out, we do still need to deploy the required models alongside the modules, though. So even if you can use the same `weaviate` container for all cases, you will still need to spin up e.g. the `text2vec-contextionary-model:en0.16.0-v...` container in your docker-compose if you want to use the `text2vec-contextionary` module.

**Comment by etiennedi** [comment]
Already implemented in `v1.2.0` - closing.
